src/core/ddsc/src/dds_read.c

```diff
@@ -554,3 +569,94 @@ dds_return_t dds_return_reader_loan (dds_reader *rd, void **buf, int32_t bufsz)
   ddsrt_mutex_unlock (&rd->m_entity.m_mutex);
   return DDS_RETCODE_OK;
 }
+
+
+void dds_transfer_samples_from_iox_to_rhc (dds_reader * rd)
+{
+#if DDS_HAS_SHM
+  void* chunk = NULL;
+  struct ddsi_domaingv* gv = rd->m_rd->e.gv;
+  thread_state_awake(lookup_thread_state(), gv);
+
+  while (true)
+  {
+    shm_lock_iox_sub(rd->m_iox_sub);
```

Comment:
It may also worth looking into whether it pays of to lock for the whole while loop (longer critical section but less lock/unlock).